rule.py

```diff
@@ -1,73 +1,154 @@
-import re 
-import sys 
-
-print("Hello") 
-print("How are you")
+from piece import Piece  
 
 ''' 
 Ruleset for Xiangqi chesspiece: soldier, cannon, chariot, horse, elephant, advisor, general 
 ''' 
+
 class Pos(object):
+
     '''
     A class for storing position of chess piece
     ''' 
+
     def __init__(self, x, y):
         self.x = x
         self.y = y 
     
 class ruleset(object): 
-    def __init__(self):
+
+    '''
+    ## Setting down ruleset for chesspieces' movements
+    ### Movements of pieces
+    Cannon: straight line movement with maximum step of 10 or 9 according to respective axes
+    Chariot: straight line movement with maximum step of 10 or 9 according to respective axes
+    Horse: L shape movement with the longer stroke takes 2 unit, short stroke takes 1 unit.
+    Elephant: Diagonal jump of 2 on x axis and 2 on y axis (pos.x +/- 2, pos.y +/- 2) with pos if the elephant position
+    Advisor: Diagonal jump of 1 both x and y axis (pos.x +/- 1, pos.y +/- 1)
+    General: Moving parallel to either axes, 1 unit each time. 
+    Soldier: Same with general, no going backwards respectively to its general.
+    
+    ### Interfering rule: Define how a movement of a chesspiece cannot be carried out.
+    Canon: another piece in its straight pathway 
+    Chariot: another piece in its straight pathway 
+    Horse: A piece placed at the middle of the longer stroke 
+    Elephant: A piece placed at the middle of the diagonal jump (pos.x +/- 1, pos.y +/- 1) with pos is the elephant position
+    
+
+    '''
+
+    def __init__(self, board_map_2d):
+        '''
+        Define domains of upper area (AREA 1 - Upper half), lower area (AREA 2 - Lower half); 
+        AREA 1 palace - Northern palace, AREA 2 palace - Southern palace;
+        Place-holder for empty position: _IS_EMPTY 
+
+        @param board_map_2d 
+        '''
         self._DOMAIN_X = 8
         self._DOMAIN_Y = 9
+        self._DOMAIN_X_UPPER_HALF = self._DOMAIN_X_LOWER_HALF = [0, 8]
+        self._DOMAIN_X_NORTHERN_PALACE = self._DOMAIN_X_SOUTHERN_PALACE = [3, 5] 
+        self._DOMAIN_Y_UPPER_HALF = [0, 4]         
+        self._DOMAIN_Y_LOWER_HALF = [5, 9]
+        self._DOMAIN_Y_NORTHERN_PALACE = [0, 2]
+        self._DOMAIN_Y_SOUTHERN_PALACE = [7, 9]
+
+        self._IS_EMPTY = '  .  '
+        self.board_map_2d = board_map_2d
     
     def is_in_board(self, new_pos):
-        if ( pos.x >= 0 and pos.x <= 8)
-            and ( pos.y >= 0 and pos.y <= 9):
+        if ( pos.x >= 0 and pos.x <= 8) and ( pos.y >= 0 and pos.y <= 9):
```

Comment:
There is no variable name pos. Maybe is should be new_pos ?

Reply:
This error occurs in a lot of function in this class. Maybe you might want to look over all of them.